designs/odcr.md

```diff
@@ -0,0 +1,121 @@
+  name: example-node-class
+spec:
+  capacityReservationSpec:
+    capacityReservationPreference: open | none | None  # Cannot be defined if capacityReservationTarget is specified
+    capacityReservationTarget: # Cannot be defined if capacityReservationPreference is specified
+      capacityReservationId: String | None
+      capacityReservationResourceGroupArn: String | None
+```
+This follows exactly how LaunchTemplate defines [CapacityReservationSpecification](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-launchtemplate-capacityreservationspecification.html) with the same
+constraints. 
+
+Karpenter will perform validation against the spec to ensure there isn't any violation prior to creating the LaunchTemplates.
+
+### Launching Nodes into Capacity Reservation
+
+#### All Launch Templates are associated with the specified Capacity Reservation
+
+EC2NodeClass currently supports automatically generating Launch Templates based on instance types and their AMI requirements however as the first iteration of supporting Capacity Reservation, we won't introduce any prioritization nor fallback nor safeguard when launching nodes against EC2NodeClasses with a Capacity Reservation. This means that it is the
+user's responsibility to ensure the Capacity Reservation associated to a EC2NodeClasses will include instance types the EC2NodeClasses could launch. Otherwise the EC2NodeClasses will fail during provisioning.
+
+This also means that we will not allow a EC2NodeClass be able to create Spot instances when user specified a Capacity Reservation. 
+
+We will skip [checkODFallback](https://github.com/aws/karpenter-provider-aws/blob/main/pkg/providers/instance/instance.go#L200C14-L200C29).
+
+_Note that these aren't permanent restrictions but simply narrowing down what features exist in the first iteration of supporting Capacity Reservation_ 
+
+Pros: 
+- This puts the onus of checking on the user. It doesn't require checking if a Capacity Reservation satisfy the requirements of an instance type by assuming the users is doing this. It makes no assumption about what a Capacity Reservation can do nor implies can do. This is helpful because if there is expectation that Karpenter does this we may be asked to
+support checking if there is capacity in the reservation before launching nodes which would significantly increase the scope of this design.
+
+Cons:
+- This implementation is overly restrictive causing it to be difficult to use. Its possible that the restrictions makes it too hard for users to use EC2NodeClasses effectively.
+
+Both a pro and a con:
+- This forces users who wish to leverage fallback and prioritization to use NodePool weights rather than relying on EC2NodeClass. I can't tell if this is a good or a bad thing but in my
+  opinion it isn't always clear what Karpenter wants users to do in this aspect.
+
+#### Labels
+
+When a node is launched against a CapacityReservation we will expose Capacity Reservation
```

Comment:
Anytime we are thinking about setting a label, we have to think if a user would select against it. Are you anticipating that users would select against the capacity reservation ID. I could perhaps see that, but I think that the reservation setting is significantly less valuable. Were you thinking of just having this there for observability information?

In either case, I think that we should callout how we expect users to use these new labels if we are going to propose them here.

Reply:
Agree on the setting, I think Id is much more helpful than setting

Reply:
If you have a dedicated NodePool for an ODCR or set of ODCRs, it does seem like you would just want to label those nodes and have a node selector on that instead of the ODCR ID.

Reply:
How do I track that my nodeclaim is part of a targeted capacity reservation that no longer exists, but I have another targeted capacity reservation that has available capacity, I need to terminate my nodeclaim and bring up a new on on the unused capacity reservation